var/spack/repos/builtin/packages/petsc/package.py

```diff
@@ -114,6 +114,7 @@ class Petsc(Package):
     depends_on('metis@5:+int64', when='+metis+int64~double')
 
     depends_on('hdf5+mpi', when='+hdf5+mpi')
+    depends_on('zlib', when='+hdf5+mpi+zlib')
```

Comment:
there is no `zlib` variant in `petsc` so I don't see how `+zlib` can be triggered in `when`.

probably `depends_on('zlib', when='^hdf5+mpi')` is what you need.